### PR TITLE
tests: dma/chan_blen_transfer: relocate TX and RX buffers

### LINKED ISSUE #75676: tests: dma: chan_blen_transfer: test breakage on Nucleo-H743ZI
**Describe the bug**

(follow-up issue to #75125)

After merge of #75245, the `chan_blen_transfer` test now succeeds on STM32WB0 MCUs but fails on Nucleo-H743ZI.

Relevant test log extract:
```
===================================================================                                                                       
START - test_tst_dma1_m2m_chan0_burst16                                                                                                   
Preparing DMA Controller: Name=dmamux@58025800, Chan_ID=4, BURST_LEN=2                                                                    
Starting the transfer                                                                                                                     
E: invalid source address                                                                                                                 
E: cannot configure the dmamux. 
```

The `invalid source address` message is emitted here:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/drivers/dma/dma_stm32_bdma.c#L566-L573
where `bdma_stm32_is_valid_memory_address` is:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/drivers/dma/dma_stm32_bdma.c#L462-L478

TL;DR: **the DMA driver refuses the transfer because the source is not in SRAM4**.

The regression comes from the fact that, beforehand, the `.rodata` and `.bss` segments of the test code could be relocated thanks to:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt#L10

And indeed, the Kconfig for Nucleo-H743ZI makes use of this feature:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/boards/nucleo_h743zi.conf#L5-L9

Before #75245, the test looked like this:
https://github.com/zephyrproject-rtos/zephyr/blob/4180d70439f599a3597b2e31753773cfa222ace2/tests/drivers/dma/chan_blen_transfer/src/test_dma.c#L25-L26
 - `tx_buff` is `static const` and initialized, so it goes in `.rodata` and gets relocated
 - `rx_buff` is `static`, not `const`, but uninitialized, so it goes in `.bss` and gets relocated 
After the patch, it now looks like this:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/src/test_dma.c#L25-L26
 - `tx_buff` is `static`, not `const`, and initialized, so it goes in `.data` **and is not relocated**
 - `rx_buff` is `static`, not `const`, but uninitialized, so it goes in `.bss` and gets relocated 

**Expected behavior**
Test should not fail.

**Impact**
Test breakage on Nucleo-H743ZI and potentially all boards with a Kconfig overlay setting `CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION` (i.e., `mimxrt` boards).

### PR DESCRIPTION
**Pull request by mathieuchopstm**
This PR places the chan_blen_transfer DMA test's TX and RX buffers in a separate file that gets entirely relocated. This ensures that the buffers reside in the correct memory, without breaking other things due to relocation of test data/code.

Fixes #75676 (tested on Nucleo-H743ZI) without introducing regression of #75125 (tested on Nucleo-WB09KE).

### CODE REVIEW

**Comment by mathieuchopstm** [comment]
Testing on NXP boards - which are the only other users of code relocation for this sample, and should have probably been broken by #75125 too - would be appreciated.

**Comment by erwango** [comment]
> Testing on NXP boards - which are the only other users of code relocation for this sample, and should have probably been broken by #75125 too - would be appreciated.

@hakehuang Would you be able to have a look ?

**Review by erwango — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Comment by hakehuang** [comment]
> @hakehuang Would you be able to have a look ?

@erwango let me run some regression test on this, the code seems good to me, will feedback soon

**Comment by mathieuchopstm** [comment]
@hakehuang 

> ```
> *** Booting Zephyr OS build v3.7.0-rc3-1-g6a9a715ee3f4 ***
> Running TESTSUITE dma_m2m
> ===================================================================
> START - test_tst_dma0_m2m_chan0_burst16
> Preparing DMA Controller: Name=dma-controller@400e8000, Chan_ID=0, BURST_LEN=2
> Starting the transfer
> ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/lib/libc/picolibc/libc-hooks.c:224
> E: r0/a1:  0x00000004  r1/a2:  0x000000e0  r2/a3:  0x40184000
> E: r3/a4:  0x00000004 r12/ip:  0x00000000 r14/lr:  0x60003839
> E:  xpsr:  0x21000000
> E: Faulting instruction address (r15/pc): 0x60007508
> E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
> E: Current thread: 0x80000838 (test_tst_dma0_m2m_chan0_burst16)
> E: Halting system
> ```

Trace isn't very informative since assertion was raised by `libc-hooks.c:224`:
https://github.com/zephyrproject-rtos/zephyr/blob/704000eb58919bb9d8f97cd4fdd6ef0e4835f672/lib/libc/picolibc/libc-hooks.c#L222-L226
This is invoked by picolibc `assert` if `_ASSERT_VERBOSE` is not set:
https://github.com/zephyrproject-rtos/picolibc/blob/764ef4e401a8f4c6a86ab723533841f072885a5b/newlib/libc/include/assert.h#L46-L54
I assume the kernel panic comes from `CODE_UNREACHABLE` generating an undefined instruction.

However, I do not see any use of C `assert` in the `dma_mcux_edma.c` driver.
The test uses `z_assert_true` which, after expansion, is a call to `z_zassert_`:
https://github.com/zephyrproject-rtos/zephyr/blob/704000eb58919bb9d8f97cd4fdd6ef0e4835f672/subsys/testsuite/ztest/include/zephyr/ztest_assert.h#L34-L43

Thus I am quite puzzled as to where this assertion failure comes from :thinking:.

_Could you run the tests with `_ASSERT_VERBOSE` enabled, which could shed some light as to what the problem is?_

> the mainline code is ok

Interesting.

_Could you run the tests again with the following patch applied (i.e., with code relocation disabled)?_

Maybe it was only needed because the i.MX DMA has a limitation similar to STM32WB0's that was worked around...? :man_shrugging:

https://github.com/zephyrproject-rtos/zephyr/blob/704000eb58919bb9d8f97cd4fdd6ef0e4835f672/tests/drivers/dma/chan_blen_transfer/boards/mimxrt1024_evk.conf#L1-L3
```diff
- CONFIG_CODE_DATA_RELOCATION=y 
 CONFIG_MEM_ATTR_HEAP=y 
- CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION="ITCM" 
```

Marking as DNM until this is worked out.

**Comment by hakehuang** [comment]
> Could you run the tests again with the following patch applied (i.e., with code relocation disabled)?

will try it and feedback once done

**Review by hakehuang — APPROVED** [review]
LGTM

**Review by teburd — APPROVED** [review]
